### PR TITLE
ENH: compress pdf files merging identical objects

### LINKED ISSUE #2768: `merge_page` increases file size of resulting pdf superlinearly
When adding a stamp onto a page using `merge_page` we observe an unexpected increase in file size. In our situation, we had a file with varying content (images, text, different page dimensions) that contained 1097 pages and the initial size was 90.8 MB. Adding a simple stamp onto every page made the resulting file (same number of pages) have a size of 213.5 MBs.

We systematically tested this with a minimal test setup leading to the following results:

Stamp: [stamp.pdf](https://github.com/user-attachments/files/16339008/stamp.pdf) | 1 page  | 31 KB 

| Test file    | Pages | Size before merging | Size after merging
| -------- | ------- | ------- | ------- | 
| [test-file2.pdf](https://github.com/user-attachments/files/16339009/test-file2.pdf)  | 8 pages |887 KB| 1.2 MB ~35% bigger
| [test-file3.pdf](https://github.com/user-attachments/files/16339010/test-file3.pdf) | 16 pages | 1.9 MB | 2.5 MB ~32% bigger
| [test-file4.pdf](https://github.com/user-attachments/files/16339012/test-file4.pdf) | 32 pages | 3.8 MB | 5.1 MB ~34% bigger
| [test-file5.pdf](https://github.com/user-attachments/files/16339025/test-file5.pdf) | 64 pages | 7.6 MB | 10.3 MB ~36% bigger
| test-file6.pdf  (too big to upload here) | 384 pages | 46,1 MB | 62,3 MB  ~35% bigger
| test-file7.pdf (too big to upload here) | 1536 pages | 111,1 MB | 175,3 MB ~58% bigger

The test files contain the same content set of 8 pages (1 times for test-file2 up to 192 times for test-file7).

These results indicate a superlinear growth in size with an increasing number of pages.

## Environment

Reproducible on py-pdf 3.7.14; also tested with 4.3.1. (latest) which shows the same behavior.

## Code + PDF

This is a minimal, complete example that shows the issue:

```python
from pypdf import PdfReader, PdfWriter

def main():
    writer = PdfWriter(clone_from="test-file7.pdf")
    for page in writer.pages:
        reader = PdfReader("stamp.pdf")
        stamp = reader.pages[0]
        page.merge_page(page2=stamp, over=True)
    writer.write('result.pdf')

if __name__ == "__main__":
    main()
```

### LINKED ISSUE #2794: ENH: compress PDF merging identical objects
@santhosh-sankaran reported in discussion #2728 that pdf compressed with PyPDF2 are smaller than the one generated by pypdf
Files_Generated_Compressed
After analysis, it has been observed that PyPDF2 was keeping a single copy of similar objects even if the original copy of the document has many copies of the objects (referenced by different indirect objects). 
pypdf keeps the same objects as the original document.

a new function can be proposed to use a single copy of the objects.

example of source pdf:
[Generated_wean_first_20_pages.pdf](https://github.com/user-attachments/files/16575458/tt2.pdf)

this file size could be reduced by more than 50%

### PR DESCRIPTION
**Pull request by pubpub-zz**
add compress_identical_objects()
discovered in #2728
closes #2794
closes #2768

### CODE REVIEW

File: pypdf/_writer.py [hunk 1]
```diff
@@ -371,10 +377,13 @@ def get_object(
...
         if isinstance(indirect_reference, int):
-            return self._objects[indirect_reference - 1]
-        if indirect_reference.pdf != self:
+            obj = self._objects[indirect_reference - 1]
+        elif indirect_reference.pdf != self:
             raise ValueError("pdf must be self")
-        return self._objects[indirect_reference.idnum - 1]
+        else:
+            obj = self._objects[indirect_reference.idnum - 1]
+        assert obj is not None
```
Comment (stefan6419846) [inline pypdf/_writer.py:385]: Please use a specific exception or remove.

File: pypdf/_writer.py [hunk 2]
```diff
@@ -1242,14 +1253,13 @@ def write_stream(self, stream: StreamType) -> None:
                 "It may not be written to correctly.",
                 __name__,
             )
+        # no more used :
+        # if not self._root:
+        #   self._root = self._add_object(self._root_object)
+        # self._sweep_indirect_references(self._root)
```
Comment (stefan6419846) [inline pypdf/_writer.py:1259]: Commented code does not make sense with Git.

File: pypdf/_writer.py [hunk 3]
```diff
@@ -1296,15 +1307,26 @@ def _write_pdf_structure(self, stream: StreamType) -> List[int]:
                     obj = self._encryption.encrypt_object(obj, idnum, 0)
                 obj.write_to_stream(stream)
                 stream.write(b"\nendobj\n")
-        return object_positions
-
-    def _write_xref_table(self, stream: StreamType, object_positions: List[int]) -> int:
+            else:
+                object_positions.append(-1)
+                free_objects.append(i + 1)
+        free_objects.append(0)  # add 0 to loop iaw PDF spec
```
Comment (stefan6419846) [inline pypdf/_writer.py:1313]:
```suggestion
        free_objects.append(0)  # add 0 to loop in accordance with PDF spec
```

File: pypdf/_writer.py [hunk 4]
```diff
@@ -1349,6 +1371,97 @@ def add_metadata(self, infos: Dict[str, Any]) -> None:
...
+        self,
+        remove_identicals: bool = True,
+        remove_orphans: bool = True,
+        verbose: int = -1,
+    ) -> None:
+        """
+        Parse the Pdf file and merge objects that have same harsh.
+        This will make objects common to multiple pages
+        Recommended to be used just before writing output
+
```
Comment (stefan6419846) [inline pypdf/_writer.py:1384]: Pdf -> PDF
harsh -> hash

And please add the trailing periods to the sentences.

File: pypdf/_writer.py [hunk 5]
```diff
@@ -1349,6 +1371,97 @@ def add_metadata(self, infos: Dict[str, Any]) -> None:
         assert isinstance(self._info, DictionaryObject)
         self._info.update(args)
 
+    def compress_identical_objects(
+        self,
+        remove_identicals: bool = True,
+        remove_orphans: bool = True,
+        verbose: int = -1,
```
Comment (stefan6419846) [inline pypdf/_writer.py:1378]: Do we really need the verbosity parameter?

File: pypdf/_writer.py [hunk 6]
```diff
@@ -1349,6 +1371,97 @@ def add_metadata(self, infos: Dict[str, Any]) -> None:
...
+        cpt = cpt_init
+        orphans = [True] * len(self._objects)
+        # look for similar objects
+        for idx, obj in enumerate(self._objects):
+            if obj is None:
+                continue
+            assert isinstance(obj.indirect_reference, IndirectObject)
+            h = obj.hash_value()
+            if cpt == 0:
+                print("+", end="", file=sys.stderr)  # noqa: T201
```
Comment (stefan6419846) [inline pypdf/_writer.py:1427]: This does not respect the verbosity.

File: pypdf/_writer.py [hunk 7]
```diff
@@ -1349,6 +1371,97 @@ def add_metadata(self, infos: Dict[str, Any]) -> None:
...
+        cnv_rev: Dict[IndirectObject, IndirectObject] = {}
+        for k, v in cnv.items():
+            cnv_rev.update(zip(v, (k,) * len(v)))
+        cpt = cpt_init
+
+        # replace reference to merged objects
+        for obj in self._objects:
+            if isinstance(obj, (DictionaryObject, ArrayObject)):
+                if cpt == 0:
+                    print(".", end="", file=sys.stderr)  # noqa: T201
```
Comment (stefan6419846) [inline pypdf/_writer.py:1447]: This does not respect the verbosity.

File: pypdf/_writer.py [hunk 8]
```diff
@@ -1349,6 +1371,97 @@ def add_metadata(self, infos: Dict[str, Any]) -> None:
...
+                    cpt = cpt_init
+                cpt -= 1
+                replace_in_obj(obj, cnv_rev)
+
+        # remove orphans (if applicable)
+        orphans[self.root_object.indirect_reference.idnum - 1] = False  # type: ignore
+        try:
+            orphans[self._info.indirect_reference.idnum - 1] = False  # type: ignore
+        except Exception:  # pragma: no cover
```
Comment (stefan6419846) [inline pypdf/_writer.py:1456]: Which exceptions could occur here to make this more specific?

File: pypdf/_writer.py [hunk 9]
```diff
@@ -1349,6 +1371,97 @@ def add_metadata(self, infos: Dict[str, Any]) -> None:
...
+            pass  # pragma: no cover
+        try:
+            orphans[self._ID.indirect_reference.idnum - 1] = False  # type: ignore
+        except Exception:
```
Comment (stefan6419846) [inline pypdf/_writer.py:1460]: Which exceptions could occur here to make this more specific?

File: pypdf/_writer.py (hunk 1 above)
Comment (pubpub-zz) [reply pypdf/_writer.py:385]: this is just used for mypy to clarify type. will add comment for clarification

File: pypdf/_writer.py (hunk 2 above)
Comment (pubpub-zz) [reply pypdf/_writer.py:1259]: I left it for understanding where _sweep_indirect_references was used

File: pypdf/_writer.py (hunk 5 above)
Comment (pubpub-zz) [reply pypdf/_writer.py:1378]: for big files this may help to see progress

File: pypdf/_writer.py (hunk 6 above)
Comment (pubpub-zz) [reply pypdf/_writer.py:1427]: to be removed

File: pypdf/_writer.py (hunk 7 above)
Comment (pubpub-zz) [reply pypdf/_writer.py:1447]: to be removed

File: pypdf/_writer.py [hunk 10]
```diff
@@ -393,7 +402,9 @@ def _replace_object(
             obj = obj.clone(self)
         self._objects[indirect_reference - 1] = obj
         obj.indirect_reference = IndirectObject(indirect_reference, gen, self)
-        return self._objects[indirect_reference - 1]
+
+        assert isinstance(obj, PdfObject)
```
[the remark this answers is not in the document]
Comment (pubpub-zz) [reply pypdf/_writer.py:406]: this is just used for mypy to clarify type.will add comment to explain

File: docs/user/file-size.md [hunk 11]
```diff
@@ -9,23 +9,17 @@ Some PDF documents contain the same object multiple times. For example, if an
...
 
-if reader.metadata is not None:
-    writer.add_metadata(reader.metadata)
+* `remove_identicals` enables/disables compression merging identical objects.
+* `remove_orphans` enables/disables suppression of unused objects.
 
-with open("smaller-new-file.pdf", "wb") as fp:
-    writer.write(fp)
-```
+It is recommended to apply this process just before writing to file/stream.
```
Comment (stefan6419846) [inline docs/user/file-size.md:22]:
```suggestion
It is recommended to apply this process just before writing to the file/stream.
```

File: pypdf/_writer.py [hunk 12]
```diff
@@ -156,12 +156,17 @@ def __init__(
...
     ) -> None:
         self._header = b"%PDF-1.3"
-        self._objects: List[PdfObject] = []
+        self._objects: List[Optional[PdfObject]] = []
         """The indirect objects in the PDF."""
 
-        self._idnum_hash: Dict[bytes, IndirectObject] = {}
-        """Maps hash values of indirect objects to their IndirectObject instances."""
+        """Maps hash values of indirect objects to the list of IndirectObjects.
+           This is used for compression
```
Comment (stefan6419846) [inline pypdf/_writer.py:163]:
```suggestion
           This is used for compression.
```

File: pypdf/_writer.py [hunk 13]
```diff
@@ -156,12 +156,17 @@ def __init__(
...
+        """
+        self._idnum_hash: Dict[bytes, Tuple[IndirectObject, List[IndirectObject]]] = {}
 
+        """list of translation already done.
```
Comment (stefan6419846) [inline pypdf/_writer.py:167]:
```suggestion
        """List of already translated IDs.
```

File: pypdf/_writer.py [hunk 14]
```diff
@@ -1349,6 +1370,79 @@ def add_metadata(self, infos: Dict[str, Any]) -> None:
...
         self._info.update(args)
 
+    def compress_identical_objects(
+        self,
+        remove_identicals: bool = True,
+        remove_orphans: bool = True,
+    ) -> None:
+        """
+        Parse the PDF file and merge objects that have same hash.
+        This will make objects common to multiple pages
```
Comment (stefan6419846) [inline pypdf/_writer.py:1380]:
```suggestion
        This will make objects common to multiple pages.
```

File: pypdf/_writer.py [hunk 15]
```diff
@@ -1349,6 +1370,79 @@ def add_metadata(self, infos: Dict[str, Any]) -> None:
...
+        Recommended to be used just before writing output
```
Comment (stefan6419846) [inline pypdf/_writer.py:1381]:
```suggestion
        Recommended to be used just before writing output.
```

File: pypdf/_writer.py [hunk 16]
```diff
@@ -1349,6 +1370,79 @@ def add_metadata(self, infos: Dict[str, Any]) -> None:
...
+
+        Args:
+            remove_identicals: remove of identical objects
```
Comment (stefan6419846) [inline pypdf/_writer.py:1384]:
```suggestion
            remove_identicals: Remove identical objects.
```

File: pypdf/_writer.py [hunk 17]
```diff
@@ -1349,6 +1370,79 @@ def add_metadata(self, infos: Dict[str, Any]) -> None:
...
+            remove_orphans: remove of unreferenced objects
```
Comment (stefan6419846) [inline pypdf/_writer.py:1385]:
```suggestion
            remove_orphans: Remove unreferenced objects.
```

**Review by stefan6419846 — APPROVED** [review]

**Pull request merged**